### PR TITLE
Fix collateral factor not taken into account on vault requirement checks

### LINKED ISSUE #1511: 🐛  dUSD withdraw and dToken loan ascendants in a dUSD Vault with additional collateral
#### What happened:
dToken loan request was rejected because of missing the requirement of 50% dUSD or DFI in the vault. I use [vault maxi](https://github.com/kuegi/defichain_maxi) for these transactions and talked to @kuegi already. We agree that it seems to be an issue on the node. Maxi creates two transaction which are built ascendants. One is a withdraw of dUSD in the vault and the other takes a loan of a dToken. `TakeLoan` is successor to the `withdraw`. (Single Mint Strategy) 

#### What you expected to happen:
I should be able to take the loan and withdraw simultaneously

#### How to reproduce it (as minimally and precisely as possible):
1. Open a vault
2. Add 55-60% dUSD and 40-45% dBTC / dETH 
3. Build two transactions:
   * withdraw dUSD
   * take a loan of a dToken
4. Withdraw succeeds
5. Loan fails

#### What are your environment parameters:
* [Vault Maxi](https://github.com/kuegi/defichain_maxi) Mainnet Vault 
* Ocean Client via [`ocean.defichain.com`](https://ocean.defichain.com)

#### Anything else we need to know?:
As you can see in the logs, the collateral ratio is above the 50% ratio even after the withdrawal and the loan. I think it is in the context of the dUSD $1.2 value in the vault that the calculation is wrong.

Maxi Debug Information
```
2022-10-13T23:00:36.276+02:00	INFO	collaterals:  | BTC: 0.90000000@19349.22222480->19449.22113778 x 1 | DUSD: 20931.30417782@1->1 x 1.2
2022-10-13T23:00:36.276+02:00	INFO	loans:  | WMT: 204.81447327@132.65070089->132.23195112
2022-10-13T23:00:36.280+02:00	INFO	starting with 157 (next: 157) in vault, target 152 - 154 (153) pair WMT-DUSD, minting only WMT
2022-10-13T23:00:36.281+02:00	INFO	doing consistency checks
2022-10-13T23:00:36.281+02:00	INFO	calculated values: collValue: 42531.86501570 loanValue: 27168.78343168 ratio: 156.54681456
2022-10-13T23:00:36.281+02:00	INFO	doing consistency checks
2022-10-13T23:00:36.300+02:00	INFO	calculated values: collValue: 42531.86501570 loanValue: 27168.78343168 ratio: 156.54681456
2022-10-13T23:00:36.301+02:00	INFO	got 867.9 sec left after reinvest
2022-10-13T23:00:36.301+02:00	INFO	increasing exposure 
2022-10-13T23:00:36.381+02:00	INFO	increasing by 629.82115373830065359477 USD, taking loan 2.6185@WMT, withdrawing 360.1562@DUSD
2022-10-13T23:00:37.402+02:00	INFO	Sending txId (withdraw): xtx1x with input: xprevoutx:1
2022-10-13T23:00:38.680+02:00	INFO	Sending txId (takeLoan): xtx2x with input: xtx1x:1
2022-10-13T23:00:41.760+02:00	INFO	error sending tx (TakeLoanTx: At least 50% of the minimum required collateral must be in DFI or DUSD (code 16)). retrying after 10 seconds
2022-10-13T23:00:51.760+02:00	INFO	error sending tx (TakeLoanTx: At least 50% of the minimum required collateral must be in DFI or DUSD (code 16)). retrying after 10 seconds
2022-10-13T23:01:01.733+02:00	INFO	error sending tx (TakeLoanTx: At least 50% of the minimum required collateral must be in DFI or DUSD (code 16)). retrying after 10 seconds
2022-10-13T23:01:12.000+02:00	INFO	error sending tx (TakeLoanTx: At least 50% of the minimum required collateral must be in DFI or DUSD (code 16)). retrying after 10 seconds
2022-10-13T23:01:21.741+02:00	INFO	error sending tx (TakeLoanTx: At least 50% of the minimum required collateral must be in DFI or DUSD (code 16)). retrying after 10 seconds
2022-10-13T23:01:31.761+02:00	INFO	failed to send tx even after after multiple retries (TakeLoanTx: At least 50% of the minimum required collateral must be in DFI or DUSD (code 16))
2022-10-13T23:01:31.761+02:00	ERROR	Error in script
2022-10-13T23:01:31.800+02:00	ERROR	WhaleApiException [Error]: 400 - BadRequest (/v0/mainnet/rawtx/send): TakeLoanTx: At least 50% of the minimum required collateral must be in DFI or DUSD (code 16)
                             	    at Object.raiseIfError (/var/task/index.js:1:243806)
                             	    at WhaleApiClient.<anonymous> (/var/task/index.js:1:247710)
                             	    at Generator.next (<anonymous>)
                             	    at fulfilled (/var/task/index.js:1:245314)
                             	    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
                             	  error: {
                             	    code: 400,
                             	    type: 'BadRequest',
                             	    at: 1665694891748,
                             	    message: 'TakeLoanTx: At least 50% of the minimum required collateral must be in DFI or DUSD (code 16)',
                             	    url: '/v0/mainnet/rawtx/send'
                             	  }
                             	}
```

### PR DESCRIPTION
**Pull request by hidiego**
#### What kind of PR is this?:

/kind fix

#### What this PR does / why we need it:
Collateral factor of DUSD and DFI was not taken into account for 50% restrictions on `takeloan` and `withdraw` actions after FortCanningEpilogue

#### Which issue(s) does this PR fixes?:

Fixes #1511

### CODE REVIEW

File: src/masternodes/mn_checks.cpp [hunk 1]
```diff
@@ -2979,17 +2979,25 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
...
+            }
 
-            if(tokenDUSD && col.nTokenId == tokenDUSD->first)
+            if(tokenDUSD && col.nTokenId == tokenDUSD->first){
                 totalCollateralsDUSD += col.nValue;
+                factorDUSD = token->factor;
+            }
         }
-        auto totalCollaterals = totalCollateralsDUSD + totalCollateralsDFI;
+        auto totalCollaterals = MultiplyAmounts(totalCollateralsDUSD, factorDUSD) + MultiplyAmounts(totalCollateralsDFI, factorDFI);
```
Comment (prasannavl) [inline src/masternodes/mn_checks.cpp:3000]: Will need a fork condition unfortunately. Let's do the following:

- Separate into two different lambas inside CollateralPctCheck. 
  - PreGreatWorldCollaterals, 
  - PostGreatWorldCollaterals
  
 and abstract this logic out into these functions.

Comment (hidiego) [reply src/masternodes/mn_checks.cpp:3000]: You mean pre/postGrandCentral or am I missing something?

Comment (kuegi) [reply src/masternodes/mn_checks.cpp:3000]: if this is released without a hardfork/blockheight-gate, you risk chain split if a takeLoan or withdraw happens that is allowed in the updated version, but not in the old version. So you need to ensure that all active nodes have the updated version before it becomes active -> fork condition

Comment (hidiego) [reply src/masternodes/mn_checks.cpp:3000]: Yes I understand this however, this was introduced in Fort Canning Epilogue (v2.11.1) and I guess the targeted version for the changes is the next fork Grand Central(v3) and not Fort Canning Great World (v2.10.0). So the conditions should be pre/postGrandCentral unless we are planning an intermediate fork which I am not aware of.

Comment (kuegi) [reply src/masternodes/mn_checks.cpp:3000]: sorry yes. my bad, didn't read properly. agreed that this will be fine in GrandCentral.

**Comment by mrgrauel** [comment]
Is the assumption correct that this changes the consensus so it does require a hardfork?

**Review by Bushstar — DISMISSED** [review]

**Comment by prasannavl** [comment]
Awaiting rebase to master. cc/ @hidiego

**Pull request merged**